[TRTLLM-4279] test: Add a protection test for checking trtllm custom ops

yali-arch opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive tests for custom PyTorch operators, including new quantized MoE operator tests and detailed torch backend attention validation.
  * Enhanced sharding transformation tests with separate pattern detection verification.
  * Refactored attention and MoE operator tests to use centralized reference implementations for consistency.
  * Updated existing tests to align with new export and transformation APIs.
* **New Features**
  * Introduced quantized FP8 and FP4 MoE custom operators with corresponding tests.
  * Added a pure PyTorch backend attention implementation with cache and comprehensive testing.
  * Added RMSNorm custom operators with FlashInfer, Triton, and Torch backends.
  * Enhanced AutoDeploy configuration and CLI argument handling with deep YAML merging.
  * Added modular export patch system with multiple export patches for improved model export.
  * Introduced a modular transform framework with staged, configurable graph transforms.
  * Added new graph transforms including build model, export to graph module, cleanup, RMSNorm fusion, and quantize MoE.
  * Enhanced sharding transformations with typed config objects and deferred execution.
  * Added sliding window and sinks support to Triton attention kernels.
  * Improved numpy-based reference attention implementation for testing.
* **Bug Fixes and Improvements**
  * Improved error handling and runtime checks in serialization and binding code.
  * Refined tensor conversion from numpy to support additional dtypes.
  * Simplified tensor caster and removed unused path caster.
  * Fixed attention metadata calculation to ensure correct page counts.
  * Updated launch configurations and documentation for AutoDeploy.
  * Enhanced logging and memory reporting in cache management.
  * Improved PyTorch export robustness by applying multiple patches and cleaning up graph modules.
  * Refactored and modularized sharding and fusion logic for better maintainability.
* **Documentation**
  * Expanded AutoDeploy example README with detailed usage and configuration guidance.
  * Updated blog post with revised Docker run command and performance tuning notes.
* **Chores**
  * Removed obsolete export transformations and replaced with modular export system.
  * Cleaned up imports and reorganized package initializers for auto-import and registration.
  * Updated dependencies to include omegaconf and pydantic-settings with YAML support.
  * Refactored code style and removed redundant returns in transformation libraries.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

[JIRA ticket/NVBugs ID/GitHub issue][fix/feat/doc/infra/...] \<summary of this PR\>

For example, assume I have a PR to support a new feature about cache manager for JIRA ticket TRTLLM-1000, it would be like:

[TRTLLM-1000][feat] Support a new feature about cache manager

Or I have a PR to fix a Llama3 accuracy issue:

[https://nvbugs/1234567][fix] Fix Llama3 accuracy issue
-->

## Description
It has been observed that the problem of missing fake implementations for new custom operations occurs from time to time, and such issues are often identified only after a certain period.
As a result, a new test file` tensorrt_llm/_torch/custom_ops/torch_custom_ops.py` is added to check the presence of fake implementations for each custom operation.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

yali-arch committed:
Add test_custom_ops.py: 2 tests to ensure each custom op has fake impl

Signed-off-by: Yang Li <yali@nvidia.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

yali-arch commented:
/bot run

yali-arch commented:
/bot kill

tensorrt-cicd commented:
[PR_Github #12568](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12568/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `fbdc650`

yali-arch committed:
Add fake impl for:
- finegrained_mixed_dtype_gemm
- set_stream
- record_event
- wait_event
- record_stream

Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

Linda-Stadter committed:
fix: bindings unit tests for nanobind (#6221)

Signed-off-by: Linda-Stadter <57756729+Linda-Stadter@users.noreply.github.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

danielafrimi committed:
Add register_fake for finegrained_mixed_dtype_gemm torch_op (#6255)

Signed-off-by: Daniel Afrimi <danielafrimi8@gmail.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

johncalesp committed:
[Issue 6193] Fix gemma3vl weight loader (#6233)

Signed-off-by: John Calderon <johncalesp@gmail.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

2ez4bz committed:
[feat] Enable TP and batching for PixtralVisionModel / Mistral3VLM (#6152)

Signed-off-by: William Zhang <133824995+2ez4bz@users.noreply.github.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

yuanjingx87 committed:
set NVIDIA_IMEX_CHANNELS for dlcluster slurm job only (#6234)

Signed-off-by: Yuanjing Xue <197832395+yuanjingx87@users.noreply.github.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

raayandhar committed:
[nvbug/5361223] doc: Update Llama4 deployment guide: update config & note concurrency (#6222)

Signed-off-by: raayandhar <rdhar@nvidia.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

lucaslie committed:
[AutoDeploy] merge feat/ad-2025-07-07 (#6196)

Signed-off-by: Gal Hubara Agam <96368689+galagam@users.noreply.github.com>
Signed-off-by: Neta Zmora <96238833+nzmora-nvidia@users.noreply.github.com>
Signed-off-by: Lucas Liebenwein <11156568+lucaslie@users.noreply.github.com>
Signed-off-by: nvchenghaoz <211069071+nvchenghaoz@users.noreply.github.com>
Signed-off-by: Frida Hou <201670829+Fridah-nv@users.noreply.github.com>
Signed-off-by: greg-kwasniewski1 <213329731+greg-kwasniewski1@users.noreply.github.com>
Signed-off-by: Suyog Gupta <41447211+suyoggupta@users.noreply.github.com>
Co-authored-by: Gal Hubara-Agam <96368689+galagam@users.noreply.github.com>
Co-authored-by: Neta Zmora <nzmora@nvidia.com>
Co-authored-by: nvchenghaoz <211069071+nvchenghaoz@users.noreply.github.com>
Co-authored-by: Frida Hou  <201670829+Fridah-nv@users.noreply.github.com>
Co-authored-by: Suyog Gupta <41447211+suyoggupta@users.noreply.github.com>
Co-authored-by: Grzegorz Kwasniewski <213329731+greg-kwasniewski1@users.noreply.github.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

Tabrizian committed:
[nvbugs/5401261][fix] Fix Triton backend disaggregated serving support (#6224)

Signed-off-by: Iman Tabrizian <10105175+tabrizian@users.noreply.github.com>
Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

yali-arch commented:
/bot run

tensorrt-cicd commented:
[PR_Github #12626](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12626/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12626](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12626/) [ run ] completed with state `FAILURE`

yali-arch committed:
Revert "Add fake impl for:"

This reverts commit 13215a7a6dd1dc841e0ac5b6c097138bb89015de.

Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>

yali-arch committed:
Fix test_custom_ops.py

Signed-off-by: Yang Li <56944310+yali-arch@users.noreply.github.com>
